### PR TITLE
have to_json and as_json return amount and currency json

### LINKED ISSUE #76: to_json, as_json and from_json should behave correctly
According to [the Rails docs](http://api.rubyonrails.org/classes/ActiveModel/Serializers/JSON.html):
`as_json` should return a hash of attributes (amount, currency) to be serialized, not a string
`from_json` is missing to assign attributes from a JSON object

We should not override to_json. A great explanation of the difference between as/to_json can be found here: https://github.com/ohler55/oj/issues/199#issuecomment-61465250

### PR DESCRIPTION
**Pull request by elfassy**
# Why
Fixes #76

# What
- `as_json` should return a hash of attributes (amount, currency) to be serialized, not a string
- `from_json` is missing to assign attributes from a JSON object

### CODE REVIEW

**Comment by elfassy** [comment]
closing for now in favor or completely removing these methods.

File: lib/money/money.rb [hunk 1]
```diff
@@ -211,11 +216,11 @@ def to_liquid
   end
 
   def to_json(options = {})
-    to_s
+    JSON.dump({value: value, currency: currency})
   end
 
   def as_json(*args)
-    to_s
+    to_json
```
Comment (bdewater) [inline lib/money/money.rb]: This should have the actual implementation.

File: lib/money/money.rb [hunk 2]
```diff
@@ -211,11 +216,11 @@ def to_liquid
   end
 
   def to_json(options = {})
-    to_s
+    JSON.dump({value: value, currency: currency})
```
Comment (bdewater) [inline lib/money/money.rb]: This method should be removed.

File: lib/money/money.rb (hunk 1 above)
Comment (elfassy) [reply lib/money/money.rb]: this was always wrong too, `as_json` needs to return a hash, not a string... fixed now

**Comment by elfassy** [comment]
DO NOT MERGE... until background jobs can handle the new format

**Comment by elfassy** [comment]
this is ready to merge now since https://github.com/Shopify/shopify/pull/151350 
@bdewater @sunblaze  let me know if you see any other objections. Will run tests in core before merging

File: lib/money/money.rb [hunk 3]
```diff
@@ -213,12 +218,8 @@ def to_liquid
...
   end
 
-  def to_json(options = {})
-    to_s
-  end
-
-  def as_json(*args)
-    to_s
+  def as_json(*_args)
+    { value: to_s(:amount), currency: currency.to_s }
```
Comment (sunblaze) [inline lib/money/money.rb]: Should we prefer `currency.iso_code` here instead to be a bit more clear and future proof if we want to_s to be more human readable?

**Comment by elfassy** [comment]
https://buildkite.com/shopify/shopify-branches/builds/487461#1596f5db-450d-44ed-b569-1a7e2f7ddcbf no go...

**Comment by sunblaze** [comment]
😮

**Comment by elfassy** [comment]
this is the new default behaviour in v1.0, with an option to fallback to the legacy behaviour

**Pull request closed without merging**